Fix git reset error: unable to read SHA1 files

## Background
When attempting to reset the git repository in the temporary directory, the following error occurs:

```
[DEBUG] Getting temporary directory: /tmp/git-repos/youngchingjui/issue-to-pr
[DEBUG] Directory exists after creation attempt: true
[DEBUG] Checking if .git and codebase exist in /tmp/git-repos/youngchingjui/issue-to-pr
[ERROR] Command failed: git reset --hard
error: unable to read sha1 file of lib/agents/thinker.ts (cfc89f1ad7bbf7e937afaba7b0eee606fab3c599)
error: unable to read sha1 file of middleware.ts (d593a4d1a30ff10a5a8cd4dfc02ce063169a843e)
fatal: Could not reset index file to revision 'HEAD'.
```

## Problem
The git reset command is failing because it cannot read the SHA1 files for certain files in the repository. This typically happens when:
1. The git repository is corrupted
2. There are permission issues
3. The git index is in an inconsistent state

## Proposed Solutions
1. Clean and re-clone the repository:
   - Delete the temporary directory
   - Re-clone the repository fresh
   - Ensure proper permissions are set

2. Repair the git repository:
   - Run git fsck to check repository integrity
   - Reset the index file
   - Verify all objects are accessible

## Tasks
- [ ] Add error handling for git reset failures
- [ ] Implement automatic repository cleanup when SHA1 errors occur
- [ ] Add logging to track frequency of these errors
- [ ] Consider adding a health check for the git repository state

## Impact
This issue affects the application's ability to handle GitHub events properly, as it cannot reset the working directory to a clean state.

--- a/lib/git.ts
+++ b/lib/git.ts
@@ -105,11 +105,6 @@ export async function checkoutBranchQuietly(
   return executeGitCommand(command, dir)
 }
 
-export async function resetHard(dir: string): Promise<string> {
-  const command = `git reset --hard`
-  return executeGitCommand(command, dir)
-}
-
 export async function cleanUntrackedFiles(dir: string): Promise<string> {
   const command = `git clean -fd`
   return executeGitCommand(command, dir)
@@ -130,11 +125,20 @@ export async function resetToOrigin(
 
 export async function cleanCheckout(branchName: string, dir: string) {
   try {
-    await checkoutBranchQuietly(branchName, dir)
-    await resetHard(dir)
+    // First clean any untracked files
     await cleanUntrackedFiles(dir)
+
+    // Fetch latest changes
     await fetchLatest(dir)
+
+    // Force clean the index before reset
+    await executeGitCommand("git rm --cached -r .", dir)
+
+    // Reset to origin
     await resetToOrigin(branchName, dir)
+
+    // Checkout the branch
+    await checkoutBranchQuietly(branchName, dir)
   } catch (error) {
     console.error(`[ERROR] Failed during clean checkout: ${error.message}`)
     throw error
@@ -174,4 +178,47 @@ export async function getDiff(dir: string): Promise<string> {
       return resolve(stdout)
     })
   })
+}
+
+export async function checkRepoIntegrity(dir: string): Promise<boolean> {
+  try {
+    // Run git fsck to check repository integrity
+    await executeGitCommand("git fsck", dir)
+    return true
+  } catch (error) {
+    console.error(`[ERROR] Repository integrity check failed: ${error.message}`)
+    return false
+  }
+}
+
+export async function cleanupRepo(dir: string): Promise<void> {
+  try {
+    // Remove the .git directory to force a fresh clone
+    await fs.rm(path.join(dir, ".git"), { recursive: true, force: true })
+    // Also remove any tracked files that might be corrupted
+    await fs.rm(dir, { recursive: true, force: true })
+  } catch (error) {
+    console.error(`[ERROR] Failed to cleanup repository: ${error.message}`)
+    throw error
+  }
+}
+
+export async function ensureValidRepo(
+  dir: string,
+  cloneUrl: string
+): Promise<void> {
+  const gitExists = await checkIfGitExists(dir)
+  if (!gitExists) {
+    await cloneRepo(cloneUrl, dir)
+    return
+  }
+
+  const isValid = await checkRepoIntegrity(dir)
+  if (!isValid) {
+    console.warn(
+      "[WARNING] Repository corruption detected, cleaning up and re-cloning"
+    )
+    await cleanupRepo(dir)
+    await cloneRepo(cloneUrl, dir)
+  }
 }

--- a/lib/utils-server.ts
+++ b/lib/utils-server.ts
@@ -7,7 +7,12 @@ import { JWT } from "next-auth/jwt"
 
 import { auth } from "@/auth"
 import { getLocalRepoDir } from "@/lib/fs"
-import { checkIfGitExists, cleanCheckout, cloneRepo } from "@/lib/git"
+import {
+  cleanCheckout,
+  cleanupRepo,
+  cloneRepo,
+  ensureValidRepo,
+} from "@/lib/git"
 import { redis } from "@/lib/redis"
 import { getCloneUrlWithAccessToken } from "@/lib/utils"
 
@@ -35,31 +40,49 @@ export async function setupLocalRepository({
 }: {
   repoFullName: string
   workingBranch?: string
-}) {
+}): Promise<string> {
   // Get or create a local directory to work off of
   const baseDir = await getLocalRepoDir(repoFullName)
 
-  // Check if .git and codebase exist in tempDir
-  console.debug(`[DEBUG] Checking if .git and codebase exist in ${baseDir}`)
-  const gitExists = await checkIfGitExists(baseDir)
-
-  if (!gitExists) {
-    // Clone the repo
-    console.debug(`[DEBUG] Cloning repo: ${repoFullName}`)
-
+  try {
     // TODO: Refactor for server-to-server auth
     const session = await auth()
     const token = session.token.access_token as string
     // Attach access token to cloneUrl
     const cloneUrlWithToken = getCloneUrlWithAccessToken(repoFullName, token)
 
-    await cloneRepo(cloneUrlWithToken, baseDir)
+    // Check repository state and repair if needed
+    await ensureValidRepo(baseDir, cloneUrlWithToken)
+
+    // Try clean checkout with retries
+    let retries = 3
+    while (retries > 0) {
+      try {
+        await cleanCheckout(workingBranch, baseDir)
+        break
+      } catch (error) {
+        retries--
+        if (retries === 0) {
+          console.error(
+            `[ERROR] Failed to clean checkout after retries: ${error.message}`
+          )
+          throw error
+        }
+        console.warn(
+          `[WARNING] Clean checkout failed, retrying... (${retries} attempts left)`
+        )
+        await cleanupRepo(baseDir)
+        await cloneRepo(cloneUrlWithToken, baseDir)
+      }
+    }
+
+    return baseDir
+  } catch (error) {
+    console.error(`[ERROR] Failed to setup repository: ${error.message}`)
+    // Clean up on failure
+    await cleanupRepo(baseDir)
+    throw error
   }
-
-  // And git pull to latest
-  await cleanCheckout(workingBranch, baseDir)
-
-  return baseDir
 }
 
 export async function refreshTokenWithLock(token: JWT) {
@@ -151,7 +174,10 @@ export async function refreshTokenWithLock(token: JWT) {
           ? JSON.parse(cachedToken)
           : cachedToken
       } catch (e) {
-        console.error(`Error parsing cached token for key ${tokenKey} after waiting:`, e)
+        console.error(
+          `Error parsing cached token for key ${tokenKey} after waiting:`,
+          e
+        )
         // If parsing fails, continue with retries
       }
     }

